### ISSUE TITLE
[Bug]: Gemma4ForConditionalGeneration + runtime LoRA greedy decode diverges from HuggingFace transformers on text-only prompts

### ISSUE THREAD

**Issue by fooSynaptic** [vllm-project/vllm#47026]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
# TODO: paste full output from H20 after:
#   export CUDA_VISIBLE_DEVICES=6
#   /root/miniconda3/envs/vllm-0.22-gemma4/bin/python collect_env.py

# Partial snapshot (repro machine · NVIDIA H20 96GB · Linux 5.4):
PyTorch version        : 2.11.0+cu130
vLLM version           : 0.22.0 (built from source against env torch; not pip wheel)
Transformers version   : 5.12.1
PEFT version           : 0.19.1
Python version         : 3.10.12
CUDA used in PyTorch   : 13.0
GPU model              : NVIDIA H20
NVIDIA driver version  : (paste from nvidia-smi)
```

</details>

## Describe the bug

**Summary:** For `google/gemma-4-E4B-it` with a PEFT LoRA adapter targeting **only** the text decoder (`model.language_model.layers.*.self_attn.q_proj|v_proj`), **greedy** (`temperature=0` / `do_sample=False`) generation in vLLM 0.22 (`enable_lora=True` + `LoRARequest`) produces **systematically different text** than HuggingFace `transformers` + `peft` on the **same prompt, same weights, same chat template**. The divergence is large enough to flip downstream classification decisions (not just cosmetic formatting).

This is **not** a transformers bug: HF is self-consistent and matches our training/eval canonical path. The mismatch appears **only** under vLLM.

### Setup (aligned on both sides)

| Item | Value |
|------|--------|
| Base model | `google/gemma-4-E4B-it` (`Gemma4ForConditionalGeneration`) |
| LoRA | r=16, alpha=32, targets `model.language_model.layers.\d+.self_attn.(q_proj\|v_proj)` (regex), `task_type=CAUSAL_LM` |
| Task | Pure **text** JSON classification (no images/audio) |
| HF load | `AutoModelForCausalLM.from_pretrained(..., attn_implementation="sdpa")` + `PeftModel` |
| vLLM load | `LLM(model=..., enable_lora=True, max_lora_rank=16)` + `LoRARequest` |
| Decode | Greedy only (`do_sample=False` / `temperature=0.0`) |
| Chat template | Gemma4 `apply_chat_template(..., add_generation_prompt=True)` |

### Observed at scale (n=29,600 paired samples, same ids)

We ran a held-out multilingual moderation benchmark (PH/RU/TR, 29,600 rows) with **identical prompts** and the **same LoRA checkpoint** on HF vs vLLM:

| Metric | HF (transformers+PEFT) | vLLM 0.22 + LoRA |
|--------|------------------------|------------------|
| Binary decision agreement @ threshold 0.5 | — | **83.31%** |
| Parsed `violation_score` exact match | — | **0.00%** |
| Positive rate @ 0.5 | 10.47% | **23.13%** |
| Distinct score values | 3 (e.g. 0.05, 0.88) | 9 (continuous-ish: 0.0, 0.3, 0.7, …) |

**Score distribution skew (illustrative):**

- HF: **89.5%** of rows at score `0.05`, **10.5%** at `0.88`
- vLLM: **76.3%** at `0.00`, **10.1%** at `0.70`, long tail up to `1.00`

Typical disagreement on the **same sample**: HF → `{"violation_type":"No Violation","violation_score":0.05}` vs vLLM → `{"violation_type":"negative","violation_score":0.7}`.

### Suspected implementation gaps (not proven root cause, but consistent with logs)

1. **Architecture path mismatch:** HF training/inference uses `AutoModelForCausalLM` + language-model LoRA; vLLM resolves `Gemma4ForConditionalGeneration` and logs many vision-tower layers as **ignored** for LoRA injection.
2. **Attention backend:** vLLM uses `TRITON_ATTN` for Gemma4 heterogeneous heads; HF uses `sdpa`. Greedy decode on a LoRA-finetuned JSON task is numerically brittle.
3. **Extra stop tokens in vLLM** (`<end_of_turn>`, etc.) may truncate differently than HF `eos_token_id`-only stopping (secondary; distribution gap persists even when parse succeeds).
4. **12B variant:** loading official `gemma-4-12b-it` for text-only LoRA additionally requires local workarounds (architecture relabel + stripping non-`language_model` weights) before vLLM will load — suggests missing first-class **text-only** Gemma4 path.

### Minimal repro script

Requires: `google/gemma-4-E4B-it`, a PEFT adapter directory (see LoRA config above), one GPU, vLLM ≥0.22 with Gemma4 LoRA support (PR #39560).

```python
"""Compare greedy JSON generation: HF transformers+PEFT vs vLLM runtime LoRA."""
import torch
from peft import PeftModel
from transformers import AutoModelForCausalLM, AutoTokenizer
from vllm import LLM, SamplingParams
from vllm.lora.request import LoRARequest

BASE = "google/gemma-4-E4B-it"  # or local path
ADAPTER = "/path/to/peft_adapter"  # r=16, q_proj+v_proj on language_model only

MESSAGES = [
    {"role": "user", "content": "Classify this text for policy violation. Reply JSON only.\n\nText: you are an idiot"},
]

# --- HuggingFace (canonical) ---
tok = AutoTokenizer.from_pretrained(BASE, trust_remote_code=True)
prompt = tok.apply_chat_template(MESSAGES, tokenize=False, add_generation_prompt=True)

model = AutoModelForCausalLM.from_pretrained(
    BASE, torch_dtype=torch.bfloat16, trust_remote_code=True, attn_implementation="sdpa"
).cuda().eval()
model = PeftModel.from_pretrained(model, ADAPTER).eval()

inputs = tok(prompt, return_tensors="pt").to("cuda")
with torch.inference_mode():
    out = model.generate(**inputs, max_new_tokens=96, do_sample=False)
hf_text = tok.decode(out[0, inputs["input_ids"].shape[1]:], skip_special_tokens=True)
print("HF:", hf_text)

# --- vLLM ---
llm = LLM(
    model=BASE,
    enable_lora=True,
    max_lora_rank=16,
    trust_remote_code=True,
    dtype="bfloat16",
    enforce_eager=True,
    max_model_len=2048,
)
sp = SamplingParams(temperature=0.0, max_tokens=96)  # no extra stop= for fair compare
vllm_text = llm.generate(
    [prompt],
    sp,
    lora_request=LoRARequest("lora", 1, ADAPTER),
)[0].outputs[0].text
print("vLLM:", vllm_text)

assert hf_text.strip() == vllm_text.strip(), "expected greedy parity"
```

### Results (observed vs expected)

**Expected:** Identical greedy token sequence (or byte-identical decoded text) when base weights, LoRA weights, prompt, and `max_new_tokens` match.

**Observed:** Different JSON content and scores on most samples; at scale only **83.31%** binary agreement and **0%** exact score match (see table above). vLLM does **not** crash — it returns well-formed JSON, but semantically wrong relative to HF.

```
# Example single-sample output (representative):
HF:   {"violation_type": "No Violation", "violation_score": 0.05, "reason": "..."}
vLLM: {"violation_type": "negative", "violation_score": 0.7, "reason": "..."}

# Full-scale paired eval (n=29600):
#   binary_agreement@0.5 = 0.8331
#   score_exact_match    = 0.0000
#   hf_positive_rate     = 0.1047
#   vllm_positive_rate   = 0.2313
```

### transformers vs vLLM checklist (per vLLM issue template)

- [x] Ran **transformers** first — output is stable and used as training reference.
- [x] Same bug class appears **only in vLLM** for the same weights/prompt.
- [x] Highlighted differences:
  - Model class path: `AutoModelForCausalLM` (HF) vs `Gemma4ForConditionalGeneration` (vLLM)
  - LoRA injection: PEFT on `language_model` attn vs Punica with vision layers skipped
  - Attention: `sdpa` vs `TRITON_ATTN`
  - Stopping: HF `eos_token_id` only vs optional extra `stop` strings in vLLM wrappers

### Request

1. Document the **canonical text-only inference path** for Gemma4 + LoRA (base class, which modules receive LoRA, recommended `SamplingParams`).
2. Fix greedy-decode parity with HuggingFace for language-model LoRA on text-only prompts (4B and 12B).
3. If full parity requires a dedicated `Gemma4ForCausalLM` / text-only checkpoint export, please document the supported workflow instead of users patching `config.json` and filtering safetensors locally.

Happy to provide more repro logs (`VLLM_LOGGING_LEVEL=DEBUG`), token-level first-divergence traces, or a small public adapter checkpoint if that helps.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by jeejeelee** [comment]
cc @[linitra24](https://github.com/linitra24)

**Comment by linitra24** [comment]
@fooSynaptic Sorry for the late reply. I tried to reproduce this with a language-only LoRA adapter that I trained myself for a Text-to-SQL task. In my local test set of 100 samples, HF + PEFT and vLLM runtime LoRA only diverged on a few cases.

For the mismatched cases I checked, the first different token usually appeared when the candidate tokens had very close logits/logprobs. For example, in one case HF preferred one token by only a very small margin, while vLLM produced nearly tied logprobs and selected the other token. So my current reproduction looks more like a small numerical difference around low-margin greedy decisions, rather than the larger distribution shift described here.

Would it be possible for you to share a sanitized or non-sensitive LoRA adapter checkpoint that reproduces the issue? That would help us investigate the exact runtime LoRA path and determine whether there is a Gemma4-specific issue in vLLM.

**Comment by fooSynaptic** [comment]
@linitra24 
Thanks for testing and for the follow-up.

**Update:** I debugged this further after filing this issue. I now believe the *large* mismatch in my original report (83% binary agreement / 0% exact score match on n=29,600, with vLLM positive rate ~23% vs HF ~10%) was **not** mainly small greedy tie-breaking on close logits. On **vLLM 0.22.0** specifically, I ran a triplet check (HF+PEFT vs vLLM+LoRA vs vLLM base-only, same weights and prompts):

| Comparison | Text exact | Score exact | Binary @0.5 |
|------------|-----------:|------------:|------------:|
| vLLM+LoRA vs **base** (0.22.0) | **100%** | **100%** | **100%** |
| vLLM+LoRA vs **HF** (0.22.0) | 0% | 0% | 0% |

So on 0.22.0, vLLM+LoRA output was **identical to base-only** — runtime LoRA did not appear to change generation at all. That matches the behavior discussed around **#39815 / #39816** (Gemma4 module alias → LoRA silently reset).

After I moved to a **nightly build that includes #39816** (`0.22.1rc1.dev266+`), on the same adapter and the same 29,600 prompts:

| Comparison | Score exact | Binary @0.5 |
|------------|------------:|------------:|
| vLLM+LoRA vs HF (nightly) | **99.93%** | **99.93%** |

So the original “HF discrete scores (e.g. 0.05/0.88) vs vLLM very different distribution” gap **largely goes away** once LoRA actually applies. Your Text-to-SQL repro (only a few mismatches, first diff at low logit margin) sounds closer to the **residual** behavior on a fixed build, not the 0.22.0 failure mode I hit.

**Conclusion:** I’m comfortable **closing this issue** from my side — the problem I cared about for production is resolved on nightly. I’d still note for the record that **vLLM 0.22.0 + Gemma4 runtime LoRA was effectively broken** (LoRA ≈ base), and users on 0.22.x should upgrade to a build that includes #39816 rather than treating HF↔vLLM drift as a tuning problem.

Thanks again for looking into it.
